### ISSUE TITLE
per-proc per-syscall stats (Bugzilla #1082)

### ISSUE THREAD

**Issue by pcpemail** [performancecopilot/pcp#168]
On 2014-11-20 10:39:47 -0600, Frank Ch. Eigler wrote:

> buffalo item
> 
> - pcp todo item 16: syscall stats
>   - sometimes need to be able to measure per-process per-syscall stats
>     e.g., proc.syscall.FOO ["pid"]   or proc.syscall ["pid:FOO"]
>   - naturally matrix-valued
>   - probably gather via systemtap or similar

**Comment by wcohen** [comment]
Thinking through how this would be designed and implemented to work
for longer duration monitoring and more heavily loaded systems.  The
example script syscallbypid-nd.stp
(https://sourceware.org/systemtap/examples/lwtools/syscallbypid-nd.stp)
probes and stores per-proc per-syscall stat information.  However,
there are some drawbacks with syscallbypid-nd.stp:

-high overhead: compiling, starting, stop script because of the large number of function probes it creates and installs
-script is fragile because uses every syscall in tapset, any breakage in any nd_syscall.* is noticeable
-associative arrays are likely to overlow with any significant activity on the machine


The high overhead and fragility could be addressed by using the syscall tracepoint:

kernel.trace("raw_syscalls:sys_enter") $regs:struct pt_regs* $id:long int

The $id provides the syscall number but not the syscall name.  The
mapping of the syscall number back to the syscall done something like
bcc syscount.py
https://github.com/iovisor/bcc/blob/master/tools/syscount.py using
"ausyscall --dump".

The overflow of the associative arrays is still a concern.  Over time
associative array would fill up with entries for processes.  Maybe
prune the arrays to limit the entries to processes that are currently
running.  When a process exits would need to go through and delete the
each of the entries associated with the exited process.  The pruning
could be expensive since a foreach would have to scan through the
entire associative array to find any entries for that process because
there isn't the equivalent to wildcard matching for the delete
operation.  In the mean time any other syscall trying to access the
associative array could be blocked by the foreach loop pruning entries
from the exiting process

Did an experiment to see if linear histograms could be used in
systemtap to store the syscall information, one linear histogram per
process.  Thinking this would allow a histogram be removed for a
process with a single delete rather than using a foreach loop to scan
through all the entries. Linear historgrams are limited to 128 entries
and there are more syscall than that, so using linear hitograms is not
feasible.

**Comment by wcohen** [comment]
Here is an quick, incomplete example. Looks like skips might be a
problem as this was on a lightly loaded laptop.

$ more syscallsbypid.stp
#! /usr/bin/env stap
#
# Copyright (C) 2018 Red Hat, Inc.
# By William Cohen, Red Hat Inc.
# wcohen@redhat.com
#
# 
#
# USAGE: stap syscallbypid.stp
#

global arr

probe kernel.trace("raw_syscalls:sys_enter") { arr[pid(), $id]++ }

# FIXME clean up after exiting processes
# probe scheduler.process_exit { delete arr[pid] }

#FIXME should be printing out the syscall names rather than the numbers
probe prometheus {
    @prometheus_dump_array2(arr, "count", "pid", "syscall")
}
[wcohen@cervelo profiling]$ stap syscallsbypid.stp
Missing separate debuginfos, use: debuginfo-install kernel-core-4.17.11-200.fc28.x86_64 
ERROR: Skipped too many probes, check MAXSKIPPED or try again with stap -t for more details.
WARNING: Number of errors: 0, skipped probes: 106
WARNING: /usr/bin/staprun exited with status: 1
Pass 5: run failed.  [man error::pass5]

**Comment by myllynen** [comment]
Perhaps you could also consider / investigate adding the BCC syscount tool (https://github.com/iovisor/bcc/blob/master/tools/syscount_example.txt) as a PCP BCC PMDA module.

**Comment by fche** [comment]
Will, the skipped probes aren't really a big problem.  A busy machine will have to tolerate missed counts and filled-up arrays.
Deleting records for a dead process is a matter of  `delete arr[pid(),*]`.

    global arr%[20000]
    probe kernel.trace("sys_enter") { arr[pid(),$id] ++ }
    probe kernel.trace("sched_process_exit") { delete arr[pid(),*] }
    probe prometheus.{ @prometheus_dump_array2(arr, "count", "pid", "syscall") }

The only thing it lacks is a mapping from syscall# to syscall name, which would be assisted by the `syscall_name()` function, which a variant of the @prometheus_dump_array2 macro would have to call on the values as it goes along.

**Comment by wcohen** [comment]
The bcc syscount.py is aggregating the syscalls across all processes.  The sticking point is the per pid accounting of the information.  Having a bpf map or a systemtap associative array accumulate each syscall across all processes is straight forward. Having to do this on each process can quickly exceed the maximum number of entries for those data structures.  Using "global arr%[20000]" suggested by Frank addresses the map overflow issue.  It will have interesting effects on the counts produced as counts could drop as an pid/syscall entry is elimated by another pid/syscall count then later returns with a lower count.

The skips are still an issue.  Is there some way to disable that?  Otherwise once 100 skips is reached the script exits.

There is an equivalent tapset point scheduler.process_exit for kernel.trace("sched_process_exit"), but there doesn't seem to be a tapset point for kernel.trace("sys_enter").

The information for the syscall numbers can be generated via "ausyscall --dump".  Still need to figure where to hook that in .  Would prefer not to do it in the kernel.trace("sys_enter") handler to avoid slowing down syscalls.

**Comment by fche** [comment]
> The skips are still an issue. Is there some way to disable that?

The warning gives the hint: -DMAXSKIPPED, or related options like `--suppress-handler-errors`.

> there doesn't seem to be a tapset point for kernel.trace("sys_enter")

That's ok, there's no obligation to use a tapset for such things.

> The information for the syscall numbers can be generated [...]

Systemtap already offers syscall number-to-name mapping function: The `syscall_name()` function was not hypothetical.

**Comment by wcohen** [comment]
Have syscallsbypid.stp added to the systemtap examples.

**Comment by myllynen** [comment]
Patch to implement this as BCC PMDA module sent to the list:

https://groups.io/g/pcp/topic/pmdabcc_add_module_to_trace/24501565

**Comment by fche** [comment]
(That bcc program is per-syscall, not per-process per-syscall.)

**Comment by myllynen** [comment]
I can provide per-syscall stats for the process you want to trace (so it is per-process per-syscall if you filter for one process, if the filter matches for more than one (e.g., `java`) then it'll report the combined per-syscall stats for those processes.

It cannot provide per-process per-syscall stats on every single process on the system due to BPF limitations.

It could be enhanced easily to provide per-process per-syscall for any given N processes as long as N would be small enough to keep the data structure sizes suitable for BPF.

**Comment by fche** [comment]
> due to BPF limitations.

For purposes such as this, BPF is less restrictive than you may think.  I have seen BPF-based software that used lookup tables with millions of entries.

**Comment by myllynen** [comment]
I experimented a bit more and indeed on the BPF side I was able to capture the information for thousands of processes, however the BPF program I came up with was still unable to capture everything even for the default `pid_max` (which is 32768) and in case `pid_max` is increased then the gap would be even greater. But this experiment revealed other issues:

* pmdaCache explodes due to tens or even hundreds of thousands of instances, the PMDA is taking too long time to load afterwards, and I'm pretty sure most clients would struggle to deal with such indoms
* there would be need for purging stale data (for died PIDs) which would add overhead as well
* in order to prepare for recycled PIDs the BPF data structures would also need to be cleaned (this would have to done from the BCC/PCP side since BPF does not know when a PID dies)

All in all, given the above challenges and the fact that the use case of tracing every syscall for every process separately on the system sounds quite uncommon, I'd propose we close this issue and if a need arises later have a separate issue/discussion then as needed. Thanks.

**Comment by fche** [comment]
That is fine, systemtap can supply complete per-process per-syscall data to prometheus and/or pcp now.

https://sourceware.org/git/gitweb.cgi?p=systemtap.git;a=blob;f=testsuite/systemtap.examples/profiling/syscallsbypid.stp

**Comment by myllynen** [comment]
That really doesn't solve the PCP side issues mentioned above in any way if such unfiltered information would be fed to PCP.

**Comment by wcohen** [comment]
The syscallsbypid.stp is producing prometheus formatted data.  This should be readable by PCP's Prometheus PMDA through the systemtap's  stap-exporter.

**Comment by fche** [comment]
Marko, if I read correctly, the only pcp-side issues relate to large indoms.   Yes, on workloads where there is a lot of process turnover, indoms will grow.  It's similar to the existing proc indom, just about one order of magnitude greater in terms of number of live elements.  I don't see inherent reasons why that can't perform acceptably, perhaps with further improvements to pmdaprometheus etc., or end-user client tuning of the sampling rate & choosing to log only filtered / aggregated bits of the data.

**Comment by myllynen** [comment]
It should be easy for you to test how this syscallsbypid.stp + stap-exporter + pmdaprometheus approach works by setting it up, running `while true ; ls / > /dev/null ; done`, and checking the results periodically with a PCP client.

**Comment by myllynen** [comment]
Ok, after checking the stp script in detail I see you clean the internal array on process exit so the above test is not catching most of the `ls` commands in this case.

**Comment by fche** [comment]
(Plus it would not be hard to change the stap script to retain data for expired processes for a limited holdover period.)

**Comment by myllynen** [comment]
There was after all some on and off follow-up discuss so hopefully providing the final summary here:

* pmdaCache could be eliminated (it's now possible also with BCC PMDA (see commit aab1efa) but there seems to be some Python PMDA issues around this, see issue #558), however while it speeds up PMDA startup, it doesn't solve other issues, see below
* without pmdaCache I saw occasional libpcp/cpmda.cpython/pminfo crashes with insanely large indoms (50k-100+k instances) but I don't have a reproducer for these, and tools started to slow down
* in case of tracing every single PID on the system, both BPF and stap based solutions seen so far would at some point (e.g., a script gone wild or a fork bomb) start overwriting entries for other PIDs meaning that what would be reported could be incomplete or even misleading information in some cases

Thus, I've now updated the BCC PMDA `syscount` module slightly, one can configure it with `details = True` in which case it'll also report per-pid per-syscall statistics for each traced process (in addition to the earlier provided "combined" view), with a hard-coded maximum number of PIDs to trace so that no data overwriting / information loss should be possible to occur. Perhaps this covers all the more realistic use cases, if needed we can consider additional enhancements if a real need arises in the future.

```
# pminfo -f bcc             

bcc.proc.syscall.count
    inst [1 or "write"] value 4
    inst [9 or "mmap"] value 155
    inst [10 or "mprotect"] value 87
    inst [14 or "rt_sigprocmask"] value 313
    inst [28 or "madvise"] value 76
    inst [39 or "getpid"] value 79
    inst [56 or "clone"] value 79
    inst [60 or "exit"] value 76
    inst [98 or "getrusage"] value 53
    inst [157 or "prctl"] value 79
    inst [186 or "gettid"] value 85
    inst [202 or "futex"] value 188158
    inst [204 or "sched_getaffinity"] value 158
    inst [273 or "set_robust_list"] value 79
    inst [493 or "3429::getrusage"] value 26
    inst [494 or "3429::futex"] value 14787
    inst [495 or "3429::mprotect"] value 8
    inst [496 or "3429::gettid"] value 6
    inst [498 or "3429::write"] value 4
    inst [500 or "3459::clone"] value 79
    inst [501 or "3459::getpid"] value 79
    inst [502 or "3459::getrusage"] value 27
    inst [504 or "3459::prctl"] value 79
    inst [505 or "3459::sched_getaffinity"] value 158
    inst [506 or "3459::set_robust_list"] value 79
    inst [508 or "3459::mmap"] value 155
    inst [509 or "3459::gettid"] value 79
    inst [510 or "3459::mprotect"] value 79
    inst [514 or "3459::futex"] value 173418
    inst [515 or "3459::rt_sigprocmask"] value 313
    inst [517 or "3459::exit"] value 76
    inst [518 or "3459::madvise"] value 76
# pminfo -f bcc | grep getrusage
    inst [98 or "getrusage"] value 53
    inst [493 or "3429::getrusage"] value 26
    inst [502 or "3459::getrusage"] value 27
```

Thanks.

**Comment by myllynen** [comment]
For the record, there's now (experimental) support for average and cumulative combined and per-pid per-syscall latencies in the BCC PMDA `syscount` module, see commit 9747641 for more details.

**Comment by natoscott** [comment]
@myllynen this looks like a second pmdabcc situation where one metric would be better modelled as two, with different instance domains (one per-process syscall metric, one metric with null indom for the totals) - the other situation being the recent biotop case with "pid zero".  Interestingly, both cases involve "just" allowing some BCC metrics within a cluster/module to have null indom - which may be simpler to implement from a quick look at the code, compared to allowing metrics to have arbitrary indoms within a cluster.

Also, related to that earlier labels discussion and your specific question about instance labels - this bcc.proc.syscall.count metric shows the kind of situation where we would definitely want to use instance labels (as is done in pmdaproc already in this situation).  The pid::syscall indom is a compound indom and having distinct instance labels like "pid":3429 and "syscall":"getrusage" will eventually allow pmSeries API to answer queries matching up (for example) metrics for pid 3429 from pmdabcc with metrics for that pid from pmdaproc.

cheers.

**Comment by natoscott** [comment]
| both cases involve "just" allowing some BCC metrics within a cluster/module to have null indom

On second thoughts that's not the case - for the metric in this second case, we'd want two real indoms - one for syscalls and one for pids::syscalls.  Anyway, maybe its still not too tricky to add?  :)
